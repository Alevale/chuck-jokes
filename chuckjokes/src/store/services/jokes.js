export const getJokes = (amount) => {
    return fetch(`http://localhost:3000/api/v1/jokes/random/${amount || 10}`)
        .then(result => result.json())
        .then(items => {
            return items.data.value
        })
}
