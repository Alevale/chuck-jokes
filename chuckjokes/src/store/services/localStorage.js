export const setToken = (token) => {
    if (!token) {
        throw new Error('Missing token')
    }
    localStorage.setItem('CHUCK_JOKE_token', token)
}

export const getToken = () => {
    return localStorage.getItem('CHUCK_JOKE_token')
}


