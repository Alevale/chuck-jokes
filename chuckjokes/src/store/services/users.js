const parseRequest = (res) => {
    if (res.status >= 400 && res.status < 600) {
        throw new Error('Bad response from server')
    }
    return res.json()
}

export const login = (username, password) => {
    if (!username || !password) {
        throw new Error('Missing username or password')
    }

    const url = 'http://localhost:3000/login'

    return fetch(url, {
        method: 'POST',
        body: JSON.stringify({
            username,
            password
        }),
        headers: {
            'Content-Type': 'application/json'
        }
    }).then(parseRequest)
      .then(response => {
          return response
      })
      .catch(error => {
          console.error('Error:', error)
          return Promise.reject(error)
      })
}

export const getProfile = (token) => {
    if (token) {
        return fetch('http://localhost:3000/api/v1/users/myprofile', {
            headers: {
                'Authorization': `Bearer ${token}`
            }
        })
            .then(result => result.json())
            .then(res => {
                return res
            })
    }
}

export const getUserCache = (token) => {
    if (token) {
        return fetch('http://localhost:3000/cache/get', {
            headers: {
                'Authorization': `Bearer ${token}`
            }
        })
            .then(result => result.json())
            .then(items => {
                return items
            })
    }
}

export const setUserCache = (token, state) => {
    if (token) {
        const url = 'http://localhost:3000/cache/set'

        fetch(url, {
            method: 'POST',
            body: JSON.stringify(state),
            headers: {
                'Authorization': `Bearer ${token}`,
                'Content-Type': 'application/json'
            }
        })
    }
}
