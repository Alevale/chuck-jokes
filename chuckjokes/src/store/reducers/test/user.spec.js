import reducer, {loginUser} from '../user'

import {login} from '../../services/users'

jest.mock('../../services/users', () => ({
    login: jest.fn()
}))
login.mockImplementation(() => Promise.resolve({ token: 'token', user: 'user' }))

describe('User Reducer', () => {

    it('returns an initial state when passed undefined', () => {
        const result = reducer(undefined, { type: 'ANYTHING' })
        expect(result).toMatchSnapshot()
    })

    it('it calls the login function and initiaties the application', async (done) => {
        const dispatch = jest.fn()
        await loginUser('a', 'a')(dispatch)
        expect(dispatch).toHaveBeenCalledTimes(4)
        expect(dispatch).toMatchSnapshot()
        done()
    })

})
