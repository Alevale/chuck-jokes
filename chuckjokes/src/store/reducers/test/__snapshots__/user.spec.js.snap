// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Reducer it calls the login function and initiaties the application 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "payload": "token",
        "type": "SET_TOKEN_LOCALSTORAGE",
      },
    ],
    Array [
      Object {
        "payload": "user",
        "type": "LOGIN",
      },
    ],
    Array [
      [Function],
    ],
    Array [
      Object {
        "type": "LOGIN_OK",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`User Reducer returns an initial state when passed undefined 1`] = `
Object {
  "appUser": Object {},
  "finishLoading": false,
  "loading": true,
  "loginFailed": false,
  "token": undefined,
}
`;
